tm2/pkg/sdk/baseapp.go

```diff
@@ -586,6 +593,11 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliv
 		res.ResponseBase = result.ResponseBase
 		res.GasWanted = result.GasWanted
 		res.GasUsed = result.GasUsed
+
+		// This is an arbitrary value that is increased after each transaction so that each transaction in the same block
+		// retrieve a unique time when calling `time.Now()`.
+		app.txCurrentTimeOffset += 100
```

Comment:
If we change this arbitrary value of 100 in production, it will cause backward compatibility issues for contracts that depend on the timestamp value. Ideally, the offset should come from the block itself. Also, what about multiple messages per transaction?

Reply:
> what about multiple messages per transaction?

Our use case in the social app is calling a realm function like PostMessage which adds the message with its timestamp from `time.Now()`. I think it's reasonable that `time.Now()` returns the same value for each call in the _same_ transaction. It think it is OK to explain to devs that a single transaction runs "instantaneously". The only thing we're asking for in our use case is that a transaction which is submitted at a different time has a different timestamp. (Maybe the transactions are collected into one block but this is a processing detail that the devs shouldn't have to worry about.)

Reply:
Yes, the intention here is to keep the same timestamp per transaction so each message has the same and each transaction timestamp in the block is different.  We could store the offset in the block; I need to think about this more.

Reply:
I agree with @piux2 -- I'm heavily leaning towards having this be a block-level thing, rather than something that's injected through the app. We can most probably keep this kind of logic in the execution context (we as the callers of the VM control the execution context).

This way we eliminate the need for bloating the app with an offset, and propagating it out to all transactions (even ones that aren't GnoVM transactions)